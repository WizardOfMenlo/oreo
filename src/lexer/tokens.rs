//! Definition of all the tokens that we use

use super::error::LexicalError;
use serde::Serialize;

/// The main token type that we work on
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Token<'a> {
    /// Any of the keywords (apart from and and or)
    Keyword(Keyword),

    /// (, ), ;, ,
    Punctuation(Punctuation),

    /// Addition, mult, ...
    Operator(Operator),

    /// A constant
    Literal(Literal<'a>),

    /// An identifier
    Identifier(&'a str),

    /// A comment
    Comment(&'a str),

    /// Any kind of error we might encounter
    Error(LexicalError<'a>),
}

impl<'a> Token<'a> {
    /// Is the current token an error?
    pub fn is_error(&self) -> bool {
        match self {
            Token::Error(_) => true,
            _ => false,
        }
    }

    /// Is the current token a comment?
    pub fn is_comment(&self) -> bool {
        match self {
            Token::Comment(_) => true,
            _ => false,
        }
    }

    /// Is this token the same as an other (where Int(1) == Int(2))
    pub fn same_kind<'b>(&self, other: &Token<'b>) -> bool
    where
        'b: 'a,
    {
        // This should match every singleton
        if self == other {
            true
        } else {
            // Match literals
            match (self, other) {
                (Self::Literal(Literal::Integer(_)), Self::Literal(Literal::Integer(_))) => true,
                (Self::Literal(Literal::Boolean(_)), Self::Literal(Literal::Boolean(_))) => true,
                (Self::Literal(Literal::String(_)), Self::Literal(Literal::String(_))) => true,
                (Self::Identifier(_), Self::Identifier(_)) => true,
                (Self::Comment(_), Self::Comment(_)) => true,
                _ => false,
            }
        }
    }
}

/// The Keywords we have
#[allow(missing_docs)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Keyword {
    Program,
    Begin,
    End,
    Var,
    Print,
    Println,
    Get,
    While,
    If,
    Then,
    Else,
    Procedure,
    Return,
}

/// Possible Punctuation marks
#[allow(missing_docs)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Punctuation {
    Comma,
    Semicolon,
    BracketOpen,
    BracketClose,
}

/// All the operators (including boolean ones)
#[allow(missing_docs)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Operator {
    Assignement,
    Plus,
    Minus,
    Times,
    Divide,
    LesserThan,
    GreaterThan,
    Equals,
    GreaterOrEquals,
    LesserOrEquals,
    Or,
    And,
    Not,
}

/// Constants
#[allow(missing_docs)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Literal<'a> {
    Integer(isize),
    Boolean(bool),
    String(&'a str),
}
