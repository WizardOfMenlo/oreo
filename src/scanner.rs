#[derive(Debug, PartialEq)]
pub struct Scanned<'a> {
    pub(crate) inner: &'a str,
    pub(crate) line: usize,
}

#[derive(Debug, PartialEq)]
pub enum ScannedItem<'a> {
    Str(&'a str),
    UnclosedStr(&'a str),
    Rest(&'a str),
}

pub fn scan(input: &str) -> impl Iterator<Item = Scanned> {
    input
        .lines()
        .enumerate()
        .map(|(i, line)| {
            line.split_whitespace().map(move |scan| Scanned {
                inner: scan,
                line: i,
            })
        })
        .flatten()
}

struct ScanLineIterator<'a> {
    input: &'a str,
    current_position: usize,
}

impl<'a> Iterator for ScanLineIterator<'a> {
    type Item = ScannedItem<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        let remain = &self.input[self.current_position..];
        if remain.len() == 0 {
            return None;
        }

        let mut chars = remain.chars();
        let sentinel = chars.next().unwrap();

        if sentinel != '"' {
            let next_index = remain.find('"').unwrap_or_else(|| self.input.len());
            return Some(ScannedItem::Rest(&remain[0..next_index]));
        } else {
            // Find the index of the next " (No escapes atm);
            let next_index = &remain[1..].find('"');
            match next_index {
                // TODO: Here we might want to keep the "
                Some(num) => Some(ScannedItem::Str(&remain[1..*num])),
                None => Some(ScannedItem::UnclosedStr(&remain[1..])),
            }
        }
    }
}

fn scan_line(input: &str) -> impl Iterator<Item = ScannedItem> {
    ScanLineIterator {
        input,
        current_position: 0,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn scan_empty() {
        let mut res = scan("");
        assert_eq!(res.next(), None);
    }

    #[test]
    fn scan_example() {
        let res: Vec<_> = scan("(x <= > >= < y 99.88l8 )").collect();
        assert_debug_snapshot!(res);
    }

    #[test]
    fn scan_example_multiline() {
        let res: Vec<_> = scan("(x <= > >= < y 99.88l8 )\nx ) ( >= y").collect();
        assert_debug_snapshot!(res);
    }

    #[test]
    fn test_real_life() {
        let input = "program fib;\r\nbegin\r\nvar n;\r\nvar first := 0;\r\nvar second :=1;\r\nvar next;\r\nvar c :=0 ;\r\nprint \"enter the number of terms\";\r\nget n;\r\nwhile ( c < n)\r\nbegin\r\nif ( c <= 1)\r\nthen begin next := c; end\r\nelse begin\r\n next := first + second;\r\n second := next;\r\nend\r\nprint next;\r\nc := c + 1;\r\nend\r\nend\r\n";
        let res: Vec<_> = scan(input).collect();
        assert_debug_snapshot!(res);
    }
}
