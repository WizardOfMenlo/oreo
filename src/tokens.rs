#[derive(Debug, Clone, PartialEq)]
pub enum Token<'a> {
    Keyword(Keyword),
    Punctuation(Punctuation),
    Operator(Operator),
    Literal(Literal<'a>),
    Identifier(&'a str),
    Comment(&'a str),
    Error(LexicalError<'a>),
}

impl<'a> Token<'a> {
    pub fn is_error(&self) -> bool {
        match self {
            Token::Error(_) => true,
            _ => false,
        }
    }

    pub fn is_comment(&self) -> bool {
        match self {
            Token::Comment(_) => true,
            _ => false,
        }
    }

    pub fn same_kind<'b>(&self, other: &Token<'b>) -> bool
    where
        'b: 'a,
    {
        // This should match every singleton
        if self == other {
            true
        } else {
            // Match literals
            match (self, other) {
                (Self::Literal(Literal::Integer(_)), Self::Literal(Literal::Integer(_))) => true,
                (Self::Literal(Literal::Boolean(_)), Self::Literal(Literal::Boolean(_))) => true,
                (Self::Literal(Literal::String(_)), Self::Literal(Literal::String(_))) => true,
                (Self::Identifier(_), Self::Identifier(_)) => true,
                (Self::Comment(_), Self::Comment(_)) => true,
                _ => false,
            }
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Keyword {
    Program,
    Begin,
    End,
    Var,
    Print,
    Println,
    Get,
    While,
    If,
    Then,
    Else,
    Procedure,
    Return,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Punctuation {
    Comma,
    Semicolon,
    BracketOpen,
    BracketClose,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Operator {
    Assignement,
    Plus,
    Minus,
    Times,
    Divide,
    LesserThan,
    GreaterThan,
    Equals,
    GreaterOrEquals,
    LesserOrEquals,
    Or,
    And,
    Not,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Literal<'a> {
    Integer(isize),
    Boolean(bool),
    String(&'a str),
}

#[derive(Debug, Clone, PartialEq)]
pub enum LexicalError<'a> {
    ExpectedDoubleEqualsEOF,
    ExpectedAssignementEOF,
    UnclosedString(&'a str),
    UnclosedComment(&'a str),
    ExpectedDoubleEquals(char),
    ExpectedAssignement(char),
    UnknownChar(char),
}
