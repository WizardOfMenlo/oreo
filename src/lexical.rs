use crate::scanner::{Scanned, ScannedItem};
use crate::tokens::{Keyword, LexicalError, Literal, Operator, Punctuation, Token};

use lazy_static::lazy_static;
use std::collections::HashMap;

lazy_static! {
    static ref SINGLE_CHAR_TOKEN: HashMap<char, Token<'static>> = {
        let mut m = HashMap::new();
        m.insert('(', Token::Punctuation(Punctuation::BracketOpen));
        m.insert(')', Token::Punctuation(Punctuation::BracketClose));
        m.insert(';', Token::Punctuation(Punctuation::Semicolon));

        m.insert('+', Token::Operator(Operator::Plus));
        m.insert('-', Token::Operator(Operator::Minus));
        m.insert('*', Token::Operator(Operator::Times));
        m.insert('/', Token::Operator(Operator::Divide));

        m
    };
    static ref KEYWORDS: HashMap<&'static str, Token<'static>> = {
        let mut m = HashMap::new();
        m.insert("program", Token::Keyword(Keyword::Program));
        m.insert("begin", Token::Keyword(Keyword::Begin));
        m.insert("end", Token::Keyword(Keyword::End));
        m.insert("var", Token::Keyword(Keyword::Var));
        m.insert("print", Token::Keyword(Keyword::Print));
        m.insert("println", Token::Keyword(Keyword::Println));
        m.insert("get", Token::Keyword(Keyword::Get));
        m.insert("while", Token::Keyword(Keyword::While));
        m.insert("if", Token::Keyword(Keyword::If));
        m.insert("then", Token::Keyword(Keyword::Then));
        m.insert("else", Token::Keyword(Keyword::Else));

        m.insert("or", Token::Operator(Operator::Or));
        m.insert("and", Token::Operator(Operator::And));
        m.insert("not", Token::Operator(Operator::Not));
        m
    };
}

pub fn lexicalize<'a>(it: impl Iterator<Item = Scanned<'a>>) -> impl Iterator<Item = Token<'a>> {
    it.map(lexicalize_one).flatten()
}

fn lexicalize_one<'a>(scan: Scanned<'a>) -> impl Iterator<Item = Token<'a>> {
    LexicalIt::new(scan)
}

struct LexicalIt<'a> {
    s: Scanned<'a>,
    curr_position: usize,
    terminate: bool,
}

impl<'a> LexicalIt<'a> {
    fn new(s: Scanned<'a>) -> Self {
        LexicalIt {
            s,
            curr_position: 0,
            terminate: false,
        }
    }
}

impl<'a> Iterator for LexicalIt<'a> {
    type Item = Token<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.terminate {
            return None;
        }

        let remaining = match self.s.inner {
            ScannedItem::Rest(s) => &s[self.curr_position..],
            ScannedItem::Str(s) => {
                self.terminate = true;
                return Some(Token::Literal(Literal::String(s)));
            }
            ScannedItem::UnclosedStr(s) => {
                self.terminate = true;
                return Some(Token::Error(LexicalError::UnclosedString(s)));
            }
        };

        if remaining.len() == 0 {
            return None;
        }

        let mut chars = remaining.chars();
        let start_char = chars.next().unwrap();

        // This resolves all unambiguos single chars
        if SINGLE_CHAR_TOKEN.contains_key(&start_char) {
            self.curr_position += 1;
            return SINGLE_CHAR_TOKEN.get(&start_char).cloned();
        }

        // Handle the single unambiguos chars
        if start_char == '<' || start_char == '>' || start_char == ':' || start_char == '=' {
            let next = chars.next();
            let (token, move_forward) = match (start_char, next) {
                ('<', Some('=')) => (Token::Operator(Operator::LesserOrEquals), true),
                ('<', None) | ('<', Some(_)) => (Token::Operator(Operator::LesserThan), false),

                ('>', Some('=')) => (Token::Operator(Operator::GreaterOrEquals), true),
                ('>', None) | ('>', Some(_)) => (Token::Operator(Operator::GreaterThan), false),

                (':', Some('=')) => (Token::Operator(Operator::Assignement), true),
                (':', Some(x)) => (Token::Error(LexicalError::ExpectedAssignement(x)), false),
                (':', None) => (Token::Error(LexicalError::ExpectedAssignementEOF), false),

                ('=', Some('=')) => (Token::Operator(Operator::Equals), true),
                ('=', Some(x)) => (Token::Error(LexicalError::ExpectedDoubleEquals(x)), false),
                ('=', None) => (Token::Error(LexicalError::ExpectedDoubleEqualsEOF), false),

                // If we get here, a lot must be wrong
                _ => return Some(Token::Error(LexicalError::UnknownChar('?'))),
            };

            self.curr_position += 1 + move_forward as usize;
            return Some(token);
        }

        // Get a integer literal
        if start_char.is_numeric() {
            let (token, read) = parse_literal(remaining);
            self.curr_position += read;
            return Some(token);
        }

        // Handle stuff that looks like identifiers
        if start_char.is_alphabetic() || start_char == '_' {
            let (token, read) = parse_identifier(remaining);
            self.curr_position += read;
            return Some(token);
        }

        self.curr_position += 1;
        Some(Token::Error(LexicalError::UnknownChar(start_char)))
    }
}

fn parse_literal(input: &str) -> (Token, usize) {
    // ['1', '0', 'a']
    //             ^
    // ['1', '0']
    // 10
    let first_non_number = input
        .find(|c: char| !c.is_numeric())
        .unwrap_or_else(|| input.len());
    let literal = &input[0..first_non_number];
    (
        Token::Literal(Literal::Integer(literal.parse().unwrap())),
        first_non_number,
    )
}

fn parse_identifier(input: &str) -> (Token, usize) {
    let first_non_id = input
        .find(|c: char| !(c.is_alphanumeric() || c == '_'))
        .unwrap_or_else(|| input.len());
    let id = &input[0..first_non_id];

    if KEYWORDS.contains_key(id) {
        return (KEYWORDS.get(id).cloned().unwrap(), first_non_id);
    }

    (Token::Identifier(id), first_non_id)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::scanner::scan;
    use insta::assert_debug_snapshot;

    #[test]
    fn lex_empty() {
        let mut res = lexicalize(scan(""));
        assert_eq!(res.next(), None);
    }

    #[test]
    fn lex_example() {
        let res: Vec<_> = lexicalize(scan("(x <= > >= < y 99.88l8 )")).collect();
        assert_debug_snapshot!(res);
    }

    #[test]
    fn lex_real_life() {
        let input = "program fib;\r\nbegin\r\nvar n;\r\nvar first := 0;\r\nvar second :=1;\r\nvar next;\r\nvar c :=0 ;\r\nprint \"enter the number of terms\";\r\nget n;\r\nwhile ( c < n)\r\nbegin\r\nif ( c <= 1)\r\nthen begin next := c; end\r\nelse begin\r\n next := first + second;\r\n second := next;\r\nend\r\nprint next;\r\nc := c + 1;\r\nend\r\nend\r\n";
        let res: Vec<_> = lexicalize(scan(input)).collect();
        assert_debug_snapshot!(res);
    }
}
