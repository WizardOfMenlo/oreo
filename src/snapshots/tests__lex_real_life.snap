---
source: src/lexical.rs
expression: res
---
[
    EnrichedToken {
        inner: Keyword(
            Program,
        ),
        line_no: 0,
    },
    EnrichedToken {
        inner: Identifier(
            "fib",
        ),
        line_no: 0,
    },
    EnrichedToken {
        inner: Punctuation(
            Semicolon,
        ),
        line_no: 0,
    },
    EnrichedToken {
        inner: Keyword(
            Begin,
        ),
        line_no: 1,
    },
    EnrichedToken {
        inner: Keyword(
            Var,
        ),
        line_no: 2,
    },
    EnrichedToken {
        inner: Identifier(
            "n",
        ),
        line_no: 2,
    },
    EnrichedToken {
        inner: Punctuation(
            Semicolon,
        ),
        line_no: 2,
    },
    EnrichedToken {
        inner: Keyword(
            Var,
        ),
        line_no: 3,
    },
    EnrichedToken {
        inner: Identifier(
            "first",
        ),
        line_no: 3,
    },
    EnrichedToken {
        inner: Operator(
            Assignement,
        ),
        line_no: 3,
    },
    EnrichedToken {
        inner: Literal(
            Integer(
                0,
            ),
        ),
        line_no: 3,
    },
    EnrichedToken {
        inner: Punctuation(
            Semicolon,
        ),
        line_no: 3,
    },
    EnrichedToken {
        inner: Keyword(
            Var,
        ),
        line_no: 4,
    },
    EnrichedToken {
        inner: Identifier(
            "second",
        ),
        line_no: 4,
    },
    EnrichedToken {
        inner: Operator(
            Assignement,
        ),
        line_no: 4,
    },
    EnrichedToken {
        inner: Literal(
            Integer(
                1,
            ),
        ),
        line_no: 4,
    },
    EnrichedToken {
        inner: Punctuation(
            Semicolon,
        ),
        line_no: 4,
    },
    EnrichedToken {
        inner: Keyword(
            Var,
        ),
        line_no: 5,
    },
    EnrichedToken {
        inner: Identifier(
            "next",
        ),
        line_no: 5,
    },
    EnrichedToken {
        inner: Punctuation(
            Semicolon,
        ),
        line_no: 5,
    },
    EnrichedToken {
        inner: Keyword(
            Var,
        ),
        line_no: 6,
    },
    EnrichedToken {
        inner: Identifier(
            "c",
        ),
        line_no: 6,
    },
    EnrichedToken {
        inner: Operator(
            Assignement,
        ),
        line_no: 6,
    },
    EnrichedToken {
        inner: Literal(
            Integer(
                0,
            ),
        ),
        line_no: 6,
    },
    EnrichedToken {
        inner: Punctuation(
            Semicolon,
        ),
        line_no: 6,
    },
    EnrichedToken {
        inner: Keyword(
            Print,
        ),
        line_no: 7,
    },
    EnrichedToken {
        inner: Literal(
            String(
                "enter the number of terms",
            ),
        ),
        line_no: 7,
    },
    EnrichedToken {
        inner: Punctuation(
            Semicolon,
        ),
        line_no: 7,
    },
    EnrichedToken {
        inner: Keyword(
            Get,
        ),
        line_no: 8,
    },
    EnrichedToken {
        inner: Identifier(
            "n",
        ),
        line_no: 8,
    },
    EnrichedToken {
        inner: Punctuation(
            Semicolon,
        ),
        line_no: 8,
    },
    EnrichedToken {
        inner: Keyword(
            While,
        ),
        line_no: 9,
    },
    EnrichedToken {
        inner: Punctuation(
            BracketOpen,
        ),
        line_no: 9,
    },
    EnrichedToken {
        inner: Identifier(
            "c",
        ),
        line_no: 9,
    },
    EnrichedToken {
        inner: Operator(
            LesserThan,
        ),
        line_no: 9,
    },
    EnrichedToken {
        inner: Identifier(
            "n",
        ),
        line_no: 9,
    },
    EnrichedToken {
        inner: Punctuation(
            BracketClose,
        ),
        line_no: 9,
    },
    EnrichedToken {
        inner: Keyword(
            Begin,
        ),
        line_no: 10,
    },
    EnrichedToken {
        inner: Keyword(
            If,
        ),
        line_no: 11,
    },
    EnrichedToken {
        inner: Punctuation(
            BracketOpen,
        ),
        line_no: 11,
    },
    EnrichedToken {
        inner: Identifier(
            "c",
        ),
        line_no: 11,
    },
    EnrichedToken {
        inner: Operator(
            LesserOrEquals,
        ),
        line_no: 11,
    },
    EnrichedToken {
        inner: Literal(
            Integer(
                1,
            ),
        ),
        line_no: 11,
    },
    EnrichedToken {
        inner: Punctuation(
            BracketClose,
        ),
        line_no: 11,
    },
    EnrichedToken {
        inner: Keyword(
            Then,
        ),
        line_no: 12,
    },
    EnrichedToken {
        inner: Keyword(
            Begin,
        ),
        line_no: 12,
    },
    EnrichedToken {
        inner: Identifier(
            "next",
        ),
        line_no: 12,
    },
    EnrichedToken {
        inner: Operator(
            Assignement,
        ),
        line_no: 12,
    },
    EnrichedToken {
        inner: Identifier(
            "c",
        ),
        line_no: 12,
    },
    EnrichedToken {
        inner: Punctuation(
            Semicolon,
        ),
        line_no: 12,
    },
    EnrichedToken {
        inner: Keyword(
            End,
        ),
        line_no: 12,
    },
    EnrichedToken {
        inner: Keyword(
            Else,
        ),
        line_no: 13,
    },
    EnrichedToken {
        inner: Keyword(
            Begin,
        ),
        line_no: 13,
    },
    EnrichedToken {
        inner: Identifier(
            "next",
        ),
        line_no: 14,
    },
    EnrichedToken {
        inner: Operator(
            Assignement,
        ),
        line_no: 14,
    },
    EnrichedToken {
        inner: Identifier(
            "first",
        ),
        line_no: 14,
    },
    EnrichedToken {
        inner: Operator(
            Plus,
        ),
        line_no: 14,
    },
    EnrichedToken {
        inner: Identifier(
            "second",
        ),
        line_no: 14,
    },
    EnrichedToken {
        inner: Punctuation(
            Semicolon,
        ),
        line_no: 14,
    },
    EnrichedToken {
        inner: Identifier(
            "second",
        ),
        line_no: 15,
    },
    EnrichedToken {
        inner: Operator(
            Assignement,
        ),
        line_no: 15,
    },
    EnrichedToken {
        inner: Identifier(
            "next",
        ),
        line_no: 15,
    },
    EnrichedToken {
        inner: Punctuation(
            Semicolon,
        ),
        line_no: 15,
    },
    EnrichedToken {
        inner: Keyword(
            End,
        ),
        line_no: 16,
    },
    EnrichedToken {
        inner: Keyword(
            Print,
        ),
        line_no: 17,
    },
    EnrichedToken {
        inner: Identifier(
            "next",
        ),
        line_no: 17,
    },
    EnrichedToken {
        inner: Punctuation(
            Semicolon,
        ),
        line_no: 17,
    },
    EnrichedToken {
        inner: Identifier(
            "c",
        ),
        line_no: 18,
    },
    EnrichedToken {
        inner: Operator(
            Assignement,
        ),
        line_no: 18,
    },
    EnrichedToken {
        inner: Identifier(
            "c",
        ),
        line_no: 18,
    },
    EnrichedToken {
        inner: Operator(
            Plus,
        ),
        line_no: 18,
    },
    EnrichedToken {
        inner: Literal(
            Integer(
                1,
            ),
        ),
        line_no: 18,
    },
    EnrichedToken {
        inner: Punctuation(
            Semicolon,
        ),
        line_no: 18,
    },
    EnrichedToken {
        inner: Keyword(
            End,
        ),
        line_no: 19,
    },
    EnrichedToken {
        inner: Keyword(
            End,
        ),
        line_no: 20,
    },
]
